### ISSUE TITLE
Permissions on /var/www/

### ISSUE THREAD

**Issue by ochorocho** [koehnlein/docker-typo3#1]
After install i get a white page due to missing permissions on /var/www/

~~I think this can be fixed by adding `USER apache` before this block
https://github.com/koehnlein/docker-typo3/blob/master/Dockerfile.php72-alpine#L22-L24
or `RUN chown apache:apache /var/www/` after.~~

I get this error due to missing permissions for apache user:
[Fri Aug 24 08:48:12.076603 2018] [php7:notice] [pid 6] [client 172.24.0.1:52258] PHP Fatal error:  Uncaught TYPO3\\CMS\\Core\\Cache\\Exception\\NoSuchCacheException: A cache with identifier "assets" does not exist. in /var/www/html/typo3/sysext/core/Classes/Cache/CacheManager.php:131\nStack trace:\n#0 /var/www/html/typo3/sysext/core/Classes/Imaging/IconRegistry.php(559): TYPO3\\CMS\\Core\\Cache\\CacheManager->getCache('assets')\n#1 /var/www/html/typo3/sysext/core/Classes/Imaging/IconRegistry.php(533): TYPO3\\CMS\\Core\\Imaging\\IconRegistry->getCachedBackendIcons()\n#2 /var/www/html/typo3/sysext/core/Classes/Imaging/IconRegistry.php(522): TYPO3\\CMS\\Core\\Imaging\\IconRegistry->initialize()\n#3 /var/www/html/typo3/sysext/core/Classes/Utility/GeneralUtility.php(3586): TYPO3\\CMS\\Core\\Imaging\\IconRegistry->__construct()\n#4 /var/www/html/typo3/sysext/core/Classes/Imaging/IconFactory.php(65): TYPO3\\CMS\\Core\\Utility\\GeneralUtility::makeInstance('TYPO3\\\\CMS\\\\Core\\\\...')\n#5 /var/www/html/typo3/sysext/core/Classes/Utility/GeneralUtility.php(3586): TYPO3\\CMS\\Core\\Imaging\\IconFactory->__construct()\n#6 /var/www/html/typo3/sysext/core/Clas in /var/www/html/typo3/sysext/core/Classes/Cache/CacheManager.php on line 131

Any hints how to fix it?

**Comment by koehnlein** [comment]
I cannot see a reason, why apache should need to write permissions to `/var/www`. Of course, apache needs to write some files in any subdirectories. But not `/var/www` itsself?!

I googled _"A cache with identifier "assets" does not exist"_ and found https://forge.typo3.org/issues/84654 which is related with https://forge.typo3.org/issues/84646, where I found ...

> It was a permission problem for my path `typo3temp/logs`. I had manually created that, webserver did not have write permission.

I can imagine, you have the same problem. Who did create the subfolders inside your container? Which permissions do they have? Can you run `typo3cms install:fixfolderstructure`? Does it work? If not, which error message do you see?

**Comment by ochorocho** [comment]
Sorry for the delay. 

Permissions ... 

```sh
drwxr-xr-x    1 root     root          4096 Aug 27 11:23 .
drwxr-xr-x    1 root     root          4096 Aug 15 10:49 ..
drwxr-xr-x   12 root     root           408 Aug 27 11:33 html
drwxr-xr-x    4 root     root          4096 Aug 15 10:49 localhost
lrwxrwxrwx    1 root     root            16 Aug 15 10:49 logs -> /var/log/apache2
lrwxrwxrwx    1 root     root            16 Aug 15 10:49 modules -> /usr/lib/apache2
drwxr-xr-x    3 root     root           102 Aug 27 11:20 packages
lrwxrwxrwx    1 root     root            12 Aug 15 10:49 run -> /run/apache2
drwxr-xr-x    4 root     root           136 Aug 27 11:20 tests
drwxrwsr-x    4 root     root          4096 Aug 27 11:23 var
drwxr-xr-x   25 root     root           850 Aug 27 11:23 vendor
```

inside `html`

```sh
drwxr-xr-x   12 root     root           408 Aug 27 11:33 .
drwxr-xr-x    1 root     root          4096 Aug 27 11:23 ..
-rw-r--r--    1 root     root         13377 Aug 27 11:20 .htaccess
-rw-r--r--    1 root     root           332 Aug 27 11:20 .htaccess.YourEnvironmentHere
-rw-r--r--    1 root     root             0 Aug 27 11:20 FIRST_INSTALL
drwxrwsr-x    4 root     root           136 Aug 27 11:33 fileadmin
-rw-r--r--    1 root     root            53 Aug 27 11:20 google7e71f3a488fa33fd.html
-rw-r--r--    1 root     root           975 Aug 27 11:23 index.php
drwxr-xr-x    5 root     root           170 Aug 27 11:23 typo3
drwxr-xr-x    7 root     root           238 Aug 27 11:33 typo3conf
drwxrwxrwx    5 root     root           170 Aug 27 11:33 typo3temp
drwxrwsr-x    4 root     root           136 Aug 27 11:33 uploads
```

`typo3cms install:fixfolderstructure` did'nt work.

I would feel more comfy when permissions in container are set to apache e.g. `chown -R apache:apache /var/www`. But newly synced files have wron permissions

Thanks for your support.

**Comment by koehnlein** [comment]
You did not add the error message from `typo3cms install:fixfolderstructure` :-)

Back to your suggestion:
Changing the permission of `/var/www` will not have any effect. Your application files will be mounted as volumes and will have the permission of the docker user. When I would change the user via `USER apache`, there would be other problems:

```
(13)Permission denied: AH00072: make_sock: could not bind to address [::]:80
(13)Permission denied: AH00072: make_sock: could not bind to address 0.0.0.0:80
no listening sockets available, shutting down
AH00015: Unable to open logs
exited with code 1
```

So I think, the correct way would be checking, which files' permissions you need to change and investigat, why they were created with wrong permissions.

**Comment by BenjaminBeck** [comment]
Had the same issue with typo3 v9 in a custom docker image. Im my opinion typo3 should detect the wrong permissions that prevent the file creation and give a more meaningful error message.
Luckily google brought me here ..

**Comment by djnotes** [comment]
Same problem here. running typo3cms install:fixfolderstructure didn't show anything. Still getting the same error. My site is located at /var/www/html/mysite/public_html

**Comment by thommyhh** [comment]
I just came across this issue, also in a docker container but this might also happen on regular hosts. The problem for me were the permissions of `var/cache`. I call `./bin/typo3cms cache:flush` to clear caches after deployment. This needs to be executed as the webserver/php user because this writes some files in `var/cache` anyway. So `var/cache` is owned by root, if the cache clearing is executed with `root` user.

I hope this helps some people.

**Comment by cepheiVV** [comment]
I had the same issue on my AWS EC2 Instance when executing the scheduler tasks.
I fixed it this way:

I checked the permissions of my web root, which is:
`> ls -la htdocs`
`drwxrws---  6 admin    www-data 4096 Dec  9 07:20 htdocs`

The scheduler task will be executed with the user `www-data`.
So I attempted to execute that manually by switching to `www-data`.
`> su www-data`
Here I got an error: "This account is currently not available"
So I looked up how to fix that on https://geekpeek.net/this-account-is-currently-not-available-login-problems/

`> nano /etc/passwd`
and set `/bin/bash` for the user `www-data` & save & exit

then tried again to `su www-data`  << which works now
& executing the scheduler task
`> /var/www/htdocs/typo3/sysext/core/bin/typo3 scheduler:run`


Now maybe some sys admin can explain a bit more why this works... :)
